### PR TITLE
[Serve] Refactor ray.serve for compatibility 

### PR DESCRIPTION
**Pull request by simon-mo**
- Use `flask` instead of `starlette` for http framework for now
- Move example actors to a file instead of a module. 
- Add serve to travis ci

### CODE REVIEW

**Comment by simon-mo** [comment]
@robertnishihara This PR is ready for review. (I can't request a reviewer somehow)

File: .travis.yml [hunk 1]
```diff
@@ -173,6 +173,11 @@ script:
   - if [ $RAY_CI_RLLIB_AFFECTED == "1" ]; then ./ci/suppress_output python python/ray/rllib/tests/test_optimizers.py; fi
   - if [ $RAY_CI_RLLIB_AFFECTED == "1" ]; then ./ci/suppress_output python python/ray/rllib/tests/test_evaluators.py; fi
 
+  # ray serve tests
+  - if [ $RAY_CI_SERVE_AFFECTED == "1" ] && [ $RAY_CI_PY3 == "1" ]; then ./ci/suppress_output python python/ray/experimental/serve/tests/test_actors.py; fi
```
Comment (robertnishihara) [inline .travis.yml]: This doesn't actually run the test, right? You need to do
```suggestion
  - if [ $RAY_CI_SERVE_AFFECTED == "1" ] && [ $RAY_CI_PY3 == "1" ]; then ./ci/suppress_output python -m pytest -v --durations=5 --timeout=300 python/ray/experimental/serve/tests/test_actors.py; fi
```

same with the others.

**Comment by robertnishihara** [comment]
@simon-mo sorry for the delay, I'm still looking over it.

The reason for switching to flask is to support Python 3.5? Did you notice any performance changes there?

File: .travis.yml [hunk 2]
```diff
@@ -173,6 +173,11 @@ script:
...
+  - if [ $RAY_CI_SERVE_AFFECTED == "1" ] && [ $RAY_CI_PY3 == "1" ]; then ./ci/suppress_output python python/ray/experimental/serve/tests/test_deadline_router.py; fi
+  - if [ $RAY_CI_SERVE_AFFECTED == "1" ] && [ $RAY_CI_PY3 == "1" ]; then ./ci/suppress_output python python/ray/experimental/serve/tests/test_default_app.py; fi
+  
   # ray tests
   # Python3.5+ only. Otherwise we will get `SyntaxError` regardless of how we set the tester.
   - if [ $RAY_CI_PYTHON_AFFECTED == "1" ]; then python -c 'import sys;exit(sys.version_info>=(3,5))' || python -m pytest -v --durations=5 --timeout=300 python/ray/experimental/test/async_test.py; fi
```
Comment (robertnishihara) [inline .travis.yml]: Let's use the `RAY_CI_PY3` variable here.

File: ci/travis/determine_tests_to_run.py [hunk 3]
```diff
@@ -54,6 +69,10 @@ def list_changed_files(commit_range):
                 RAY_CI_RLLIB_AFFECTED = 1
                 RAY_CI_LINUX_WHEELS_AFFECTED = 1
                 RAY_CI_MACOS_WHEELS_AFFECTED = 1
+            elif changed_file.startswith("python/ray/experimental/serve"):
+                RAY_CI_SERVE_AFFECTED = 1
```
Comment (robertnishihara) [inline ci/travis/determine_tests_to_run.py]: We also need to add `RAY_CI_SERVE_AFFECTED = 1` to the `python` block, the `src` bloc, and both `else` blocks.

File: .travis.yml [hunk 4]
```diff
@@ -173,6 +173,11 @@ script:
   - if [ $RAY_CI_RLLIB_AFFECTED == "1" ]; then ./ci/suppress_output python python/ray/rllib/tests/test_optimizers.py; fi
   - if [ $RAY_CI_RLLIB_AFFECTED == "1" ]; then ./ci/suppress_output python python/ray/rllib/tests/test_evaluators.py; fi
 
+  # ray serve tests
+  - if [ $RAY_CI_SERVE_AFFECTED == "1" ] && [ $RAY_CI_PY3 == "1" ]; then ./ci/suppress_output python python/ray/experimental/serve/tests/test_actors.py; fi
+  - if [ $RAY_CI_SERVE_AFFECTED == "1" ] && [ $RAY_CI_PY3 == "1" ]; then ./ci/suppress_output python python/ray/experimental/serve/tests/test_deadline_router.py; fi
+  - if [ $RAY_CI_SERVE_AFFECTED == "1" ] && [ $RAY_CI_PY3 == "1" ]; then ./ci/suppress_output python python/ray/experimental/serve/tests/test_default_app.py; fi
```
Comment (robertnishihara) [inline .travis.yml]: I think it'd be really great to figure out why these tests are failing when we run them with one big `pytest` command. I'll look into it a little.

**Comment by robertnishihara** [comment]
@simon-mo, I just noticed that we're calling `ray.experimental.get_actor` a bunch of times in the router. This call isn't particularly lightweight, and so we should be sure to do it only once per actor.

**Comment by robertnishihara** [comment]
After creating a router (and not doing any work), and letting it loop for 30 seconds, there are about 100K actor handles that the raylet is keeping track of and 100K local objects.

**Comment by pcmoritz** [comment]
Should we use aiohttp instead of flask to avoid the blocking ray.get here? In that case we can use Ray's non-blocking async get.

**Comment by robertnishihara** [comment]
@pcmoritz are you talking about for the HTTP frontend? One benefit of flask is that it will work with Python 2 (which I realize we don't support yet here, but we could..).

**Comment by pcmoritz** [comment]
Something like this:

```
from ray.experimental import async_api

async_api.init()

async def handle(request):
    result_future = # Call actor
    result = await async_api.as_future(object_id)
    # Return jsonified result
```

**Comment by simon-mo** [comment]
Yes. Aiohttp has a good async support for python 3.5. I’ll start with that.  

I realized flask is probably not idea because it’s a threaded server.

**Comment by pcmoritz** [comment]
@robertnishihara Yes, we might want to consider only supporting python 3 until somebody asks for python 2 support (which will be EOL pretty soon).

**Comment by robertnishihara** [comment]
Yeah I agree about not worrying about python 2 at this stage for ray.serve.

**Comment by robertnishihara** [comment]
Ok, once https://github.com/ray-project/ray/pull/4838 is merged, we should be able to run all of the serve tests with a single pytest command.

**Pull request closed without merging**